### DISCUSSION TITLE
Possibility to type `meta` prop in `ColumnDef`

### DISCUSSION THREAD

**Discussion by AzzouQ** [TanStack/table#4104] [Ideas] [answered]
Maybe something like this could be envisaged ?
```jsx
declare type CoreColumnDefBase<TData extends RowData, TMeta extends unknown> = {
    columns?: ColumnDef<TData>[];
    header?: ColumnDefTemplate<ReturnType<CoreHeader<TData>['getContext']>>;
    footer?: ColumnDefTemplate<ReturnType<CoreHeader<TData>['getContext']>>;
    cell?: ColumnDefTemplate<ReturnType<CoreCell<TData>['getContext']>>;
    meta?: TMeta;
};

declare type ColumnDef<TData extends RowData, TMeta extends unknown = unknown> = CoreColumnDef<TData, TMeta> & ...;
```
Then we could use something like ` useMemo<ColumnDef<TData, TMeta>[]>()` where TMeta will be optional, if not specified, it will remain unknown, otherwise the user can type it and avoid casting every time. Feel like a little bit harsh to me tho that TMeta will be added all along your typings for just one props.

**Comment by tannerlinsley** [comment, marked as answer]
I’m working on a better way to type metas. Stay tuned.

**Reply by LinnJS** [reply]
I am loving the new API so far but man typing some stuff out has been hard. I watched your talk about the learning curve of authoring in TS and I am feeling a tad bit of that pain right now trying to write Table Generics... Would love to see a video on this in the future. I want to understand and be a TS pro so badly!

**Comment by tannerlinsley** [comment]
https://tanstack.com/table/v8/docs/api/core/column-def#meta

**Reply by marceloverdijk** [reply]
@tannerlinsley I'm still having issues with this, see https://github.com/TanStack/table/discussions/4157

**Reply by AzzouQ** [reply]
@tannerlinsley you may want to upgarde the docs to include the import in the .d.ts example. For user unfamiliar with TS, it's not always obvious.

See [this issue for example](https://github.com/TanStack/table/discussions/4157#discussioncomment-3208313)
